### PR TITLE
[13.0][ADD] project_stock: New addon

### PR DESCRIPTION
**Pull request by victoralmau**
This module adds stock moves in project tasks.

Please @pedrobaeza and @sergio-teruel can you review it?

@Tecnativa TT34554

### CODE REVIEW

File: project_stock/README.rst [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/project-13-0/project-13-0-project_stock
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/140/13.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module adds stock moves in project tasks.
```
Comment (pedrobaeza) [inline project_stock/README.rst]: This module allows to consume products directly from a project task.

File: project_stock/README.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+**Table of contents**
+
+.. contents::
```
Comment (pedrobaeza) [inline project_stock/README.rst:32]: Missing configuration section where you say about where to configure source and destination location of task moves (at project level).

File: project_stock/README.rst [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+   :local:
```
Comment (pedrobaeza) [inline project_stock/README.rst:33]: Missing usage section explaining where to see the stock moves and how to put them on the task.

File: project_stock/models/account_analytic_line.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2022 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import fields, models
+
+
+class AccountAnalyticLine(models.Model):
+    _inherit = "account.analytic.line"
+
+    task_id = fields.Many2one(comodel_name="project.task", string="Project Task",)
```
Comment (pedrobaeza) [inline project_stock/models/account_analytic_line.py]: The other name already exists

```suggestion
    stock_task_id = fields.Many2one(comodel_name="project.task", string="Project Task",)
```

File: project_stock/models/account_analytic_line.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+    def _timesheet_postprocess_values(self, values):
+        res = super()._timesheet_postprocess_values(values)
+        # Delete the changes in amount if the analytic line come from task.
```
Comment (pedrobaeza) [inline project_stock/models/account_analytic_line.py]: Why is this needed?

File: project_stock/models/project_project.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                    ("code", "=", "ps_operation"),
+                    ("warehouse_id.company_id", "=", self.env.company.id),
+                ],
+                limit=1,
+            )
+            .id
+        )
+
+    @api.model
+    def _get_default_location_src_id(self):
```
Comment (pedrobaeza) [inline project_stock/models/project_project.py]: Instead of these defaults, please convert location_id and location_dest_id into computed writable fields that react when picking type is changed.

File: project_stock/models/project_project.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+# Copyright 2022 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import api, fields, models
+
+
+class ProjectProject(models.Model):
+    _inherit = "project.project"
+
+    @api.model
+    def _get_default_picking_type(self):
```
Comment (pedrobaeza) [inline project_stock/models/project_project.py]: Do it a computed writable depending on company

File: project_stock/models/project_project.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    picking_type_id = fields.Many2one(
+        comodel_name="stock.picking.type",
+        string="Operation Type",
+        domain="[('code', '=', 'ps_operation'), ('company_id', '=', company_id)]",
+        default=_get_default_picking_type,
+        index=True,
+        check_company=True,
+    )
+    location_src_id = fields.Many2one(
```
Comment (pedrobaeza) [inline project_stock/models/project_project.py]: For following standard Odoo convention, call it `location_id` instead of `location_src_id`.

File: project_stock/models/project_task.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
+# Copyright 2022 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import _, api, fields, models
+
+
+class Task(models.Model):
+    _inherit = "project.task"
+
+    @api.model
+    def _get_default_picking_type(self):
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]: Do them computed writable fields.

File: project_stock/models/project_task.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    procurement_group_id = fields.Many2one(
+        comodel_name="procurement.group", string="Procurement Group", copy=False
+    )
+    scrap_ids = fields.One2many(
+        comodel_name="stock.scrap", inverse_name="task_id", string="Scraps"
+    )
+    scrap_count = fields.Integer(
+        compute="_compute_scrap_move_count", string="Scrap Move"
+    )
+    move_raw_ids = fields.One2many(
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]: Bad name. This is not a production order with raw and finished products.

```suggestion
    move_ids = fields.One2many(
```

File: project_stock/models/project_task.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    )
+    picking_type_id = fields.Many2one(
+        comodel_name="stock.picking.type",
+        string="Operation Type",
+        domain="[('code', '=', 'ps_operation'), ('company_id', '=', company_id)]",
+        default=_get_default_picking_type,
+        index=True,
+        check_company=True,
+    )
+    location_src_id = fields.Many2one(
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]: `location_id`

File: project_stock/models/project_task.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+        string="Source Location",
+        domain="[('usage','=','internal')]",
+        default=_get_default_location_src_id,
+        index=True,
+        check_company=True,
+    )
+    location_dest_id = fields.Many2one(
+        comodel_name="stock.location",
+        string="Destination Location",
+        domain="[('usage','=','internal')]",
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]: This one shouldn't be internal. Use the same rules as the regular picking ones.

File: project_stock/models/project_task.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        default=_get_default_location_dest_id,
+        index=True,
+        check_company=True,
+    )
+    unreserve_visible = fields.Boolean(
+        string="Allowed to Unreserve Inventory",
+        compute="_compute_unreserve_visible",
+        help="Technical field to check when we can unreserve",
+    )
+    analytic_account_id = fields.Many2one(
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]:
```suggestion
    stock_analytic_account_id = fields.Many2one(
```

And add also `stock_analytic_tag_ids`

File: project_stock/models/project_task.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+        comodel_name="account.analytic.account",
+        string="Move Analytic Account",
+        help="Move created will be assigned to this analytic account",
+    )
+    analytic_line_ids = fields.One2many(
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]:
```suggestion
    stock_analytic_line_ids = fields.One2many(
```

File: project_stock/models/project_task.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+        comodel_name="account.analytic.line",
+        inverse_name="task_id",
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]:
```suggestion
        inverse_name="stock_task_id",
```

File: project_stock/models/project_task.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+            any_quantity_done = any([m.quantity_done > 0 for m in item.move_raw_ids])
+            item.unreserve_visible = not any_quantity_done and already_reserved
+
+    @api.model
+    def _prepare_procurement_group_vals(self, values):
+        return {"name": values["name"]}
+
+    @api.model
+    def create(self, values):
+        if not values.get("procurement_group_id"):
```
Comment (pedrobaeza) [inline project_stock/models/project_task.py]: Why do you need a procurement group?

File: project_stock/models/stock_move.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import _, api, exceptions, fields, models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    task_id = fields.Many2one(
+        comodel_name="project.task",
+        string="Task for finished products",
```
Comment (pedrobaeza) [inline project_stock/models/stock_move.py]: There's no finished nor raw here. Only moves. The analogy to `mrp` module was just for one part, not literally.

File: project_stock/models/stock_move.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+                    move_line_to_unlink |= move_line
+        move_line_to_unlink.unlink()
+        return True
+
+    def _action_assign(self):
+        res = super()._action_assign()
+        for move in self.filtered(lambda x: x.task_id or x.raw_material_task_id):
+            if move.move_line_ids:
+                task = move.task_id if move.task_id else move.raw_material_task_id
+                move.move_line_ids.write({"task_id": task.id})
```
Comment (pedrobaeza) [inline project_stock/models/stock_move.py]: Why do you need to pass this field to sml?

File: project_stock/models/stock_move.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+        defaults = super().default_get(fields_list)
+        if self.env.context.get("default_raw_material_task_id"):
+            task = self.env["project.task"].browse(
+                self.env.context.get("default_raw_material_task_id")
+            )
+            defaults["group_id"] = task.procurement_group_id.id
+            defaults["reference"] = task.name
+        return defaults
+
+    def _decrease_reserved_quanity(self, quantity):
```
Comment (pedrobaeza) [inline project_stock/models/stock_move.py]: Why do you need this?

File: project_stock/models/stock_move.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+        help="Technical Field to order moves",
+    )
+
+    @api.depends("state")
+    def _compute_is_done(self):
+        for move in self:
+            move.is_done = move.state in ("done", "cancel")
+
+    @api.model
+    def default_get(self, fields_list):
```
Comment (pedrobaeza) [inline project_stock/models/stock_move.py]: Why doing this on `default_get` instead of directly pass them as values to the creation?

File: project_stock/models/stock_move.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+    )
+    is_done = fields.Boolean(
+        string="Done",
+        compute="_compute_is_done",
+        store=True,
+        help="Technical Field to order moves",
+    )
+
+    @api.depends("state")
+    def _compute_is_done(self):
```
Comment (pedrobaeza) [inline project_stock/models/stock_move.py]: Why do you need this field? The condition used for this computed field is easily reproducible on views.

File: project_stock/models/stock_picking_type.py [hunk 22]
```diff
@@ -0,0 +1,9 @@
+# Copyright 2022 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import fields, models
+
+
+class StockPickingType(models.Model):
+    _inherit = "stock.picking.type"
+
+    code = fields.Selection(selection_add=[("ps_operation", "Project stock")])
```
Comment (pedrobaeza) [inline project_stock/models/stock_picking_type.py]: Why using a special code for these picking types? Is adding any value?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: project_stock/models/stock_move.py (hunk 21 above)
Comment (victoralmau) [reply project_stock/models/stock_move.py]: It is only necessary to order records in tree view.

Comment (pedrobaeza) [reply project_stock/models/stock_move.py]: I think that feature is not needed.

File: project_stock/models/stock_move.py (hunk 18 above)
Comment (victoralmau) [reply project_stock/models/stock_move.py]: I have removed this and defined the field as compute store.

File: project_stock/models/stock_move.py (hunk 19 above)
Comment (victoralmau) [reply project_stock/models/stock_move.py]: Finally this is not necessary (tests verify it).

File: project_stock/models/stock_picking_type.py (hunk 22 above)
Comment (victoralmau) [reply project_stock/models/stock_picking_type.py]: It was added only to have the moves differentiated, but with the `pickign_type_id` field I think it is enough. It is removed.

File: project_stock/models/project_task.py (hunk 16 above)
Comment (victoralmau) [reply project_stock/models/project_task.py]: It's not really necessary, it's something extra that can be removed (is removed).

File: project_stock/models/project_task.py (hunk 12 above)
Comment (victoralmau) [reply project_stock/models/project_task.py]: I have removed domains because delivery types do not have a domain.

File: project_stock/models/account_analytic_line.py (hunk 5 above)
Comment (victoralmau) [reply project_stock/models/account_analytic_line.py]: I have verified with the tests that this is not necessary.

Comment (victoralmau) [reply project_stock/models/account_analytic_line.py]: Comment added in the function to explain the reason.

**Comment by victoralmau** [comment]
Thanks for the review, all the changes are done.

**Review by gurneyalex — APPROVED** [review]

**Review by rafaelbn — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]

**Comment by victoralmau** [comment]
Please, can you try the merge process again?
